[feature/graph-api] Spaces/Drives / Graph API support (WIP)

felix-schwarz opened the pull request:
## Description
Adds support for spaces/drives / Graph API.

## Related Issue
https://github.com/owncloud/ios-sdk/pull/92

## Screenshots (if appropriate):
Spaces list | Drive root folder view
--|--
![Simulator Screen Shot - iPhone 12 mini - 2022-04-22 at 01 10 04](https://user-images.githubusercontent.com/639669/164565782-a8638de2-c0ac-42c1-b2cf-6dd2a93bf693.png) | ![Simulator Screen Shot - iPhone 12 mini - 2022-04-22 at 01 10 12](https://user-images.githubusercontent.com/639669/164565803-be43b828-a8a2-4eeb-adf3-58dda78c909e.png)
<!---
![Simulator Screen Shot - iPhone 12 mini - 2022-03-07 at 16 36 17](https://user-images.githubusercontent.com/639669/157065871-0c4a4c69-2d5f-4236-a825-6db7fbdc1308.png) | ![Simulator Screen Shot - iPhone 12 mini - 2022-03-07 at 16 36 22](https://user-images.githubusercontent.com/639669/157065894-c4ce365f-379e-426f-8ab9-e4d826c6a812.png)
-->

## Testing
- Quick overview over known issues: https://github.com/owncloud/ios-app/blob/feature/graph-api/KNOWN_ISSUES.md
- Test servers: https://owncloud.dev/ocis/deployment/continuous_deployment/
- Default/test users: https://owncloud.dev/ocis/getting-started/demo-users/
- Local test instance (remove/adapt `OCIS_URL: "https://myServerIP:9200"` when using locally or reachable externally) with docker-compose, using HTTPS on port 9200:
```yaml
---
version: "3.7"

services:
  ocis:
    image: owncloud/ocis:latest
    ports:
      - 9200:9200
      - 9215:9215
    environment:
      OCIS_INSECURE: "true"
      OCIS_URL: "https://myServerIP:9200"
      IDM_CREATE_DEMO_USERS: "true"
      IDM_ADMIN_PASSWORD: "admin"
    restart: "no"
    entrypoint:
      - /bin/sh
    # run ocis init to initialize a configuration file with random secrets
    # it will fail on subsequent runs, because the config file already exists
    # therefore we ignore the error and then start the ocis server
    command: ["-c", "ocis init || true; ocis server"]
```

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## QA

Test plan:

Reports & improvements

- [ ] (1) Go straight to the list in case no spaces are available https://github.com/owncloud/ios-app/pull/1101#issuecomment-1155346392
- [ ] (2) White background in dark theme https://github.com/owncloud/ios-app/pull/1101#issuecomment-1155357268
- [ ] (3) Add option to set as space description in txt files https://github.com/owncloud/ios-app/pull/1101#issuecomment-1155368313
- [ ] (4) `Viewer` restrictions in the account https://github.com/owncloud/ios-app/pull/1101#issuecomment-1155409366
- [ ] (5) Footer https://github.com/owncloud/ios-app/pull/1101#issuecomment-1155410715
- [ ] (6) Space image is not correctly updated when space is disabled https://github.com/owncloud/ios-app/pull/1101#issuecomment-1156586805

Commit:
[milestone/11.8.2] Milestone 11.8.2 (#1085)

* updated version and build number for 11.8.2

* [fix/pdf-editing] Fix PDF editing (#1083)

* - update SDK with fix for enterprise#4934.

* - DisplayViewController: add .allowUpdatesUntilLocalModificationPersists to temporarily allow updates to the viewed file until the rendered item again is stored remotely and no longer a local modification
- DocumentEditingAction: when invoked from a PDFViewerViewController, pass it along to EditDocumentViewController, so it can make use of allowUpdatesUntilLocalModificationPersists
- EditDocumentViewController: set allowUpdatesUntilLocalModificationPersists of a hosting PDFViewerViewController to true when a file is edit in place and changes overwrite the original

* Added changelog entry

* Calens changelog updated

* [fix/branding-folder-color] Branding: Corporate Color as Folder Color (#1086)

* Adopt the corporate branding color to the folder color

* added changelog entry

* Calens changelog updated

* Calens changelog updated

Co-authored-by: hosy <hosy@users.noreply.github.com>

* [fix/continuous-audio-playback] Fix: Continuous Audio Playback (#1088)

* #4924 fixed continuous audio playback, which stopped, after two audio files, because of not setting the activeDisplayViewController, when starting a new audio file

* Added changelog entry

* Calens changelog updated

* fixed updating navigationbar title and buttons

* fixed scaling cover image to screen size

Co-authored-by: hosy <hosy@users.noreply.github.com>

* release notes and changelog updated

* Calens changelog updated

Co-authored-by: Felix Schwarz <fs-git@iospirit.com>
Co-authored-by: hosy <hosy@users.noreply.github.com>

michaelstingl committed:
Add In House certificate option to resign README.md (#1084)

Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

michaelstingl committed:
Update release item with new download URL (#1067)

* Update release_template.md

* Update release_template.md

Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

felix-schwarz committed:
- switch to feature/graph-api SDK branch
- FileProviderExtension: change comment to reflect drive-dependant path changes

felix-schwarz committed:
- adapt to new and updated APIs (OCPath -> OCDriveID + OCLocation)
- update SDK

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=1101) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=1101) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=1101) it.</sub>

felix-schwarz committed:
- update iOS SDK
- add ClientSpacesTableViewController as temporary solution to present and pick spaces for spaces-enabled accounts
- adapt QueryFileListTableViewController, ClientDirectoryPickerViewController, ClientQueryViewController, ClientRootViewController to support drives structurally

felix-schwarz committed:
Via SDK update:
- fix duplicate drive entries
- working eTag polling via drive list

Commit:
[milestone/11.9.0] Milestone 11.9.0 (#1071)

* new version number 11.9.0, enabled beta build and warning

* [feature/infinite-propfind] Infinite PROPFIND support (#1002)

* - ProgressIndicatorViewController : extend with Progress object and cancellation support / monitoring
- BookmarkViewController: add infinite PROPFIND prepopulation step after account creation
- update SDK to provide infinite PROPFIND support

* Configuration documentation updated

* - update SDK to gain access to streaming infinite PROPFIND
- switch from receive+parse to stream+parse for infinite PROPFIND

* - update SDK to gain access to new supportsInfinitePropfind capability
- BookmarkViewController: make prepopulation configurable, using new capability and new class setting
- ProgressIndicatorViewController: add support for new title label view

* - BookmarkViewController: fix property name typo
- scheme: add new options to test streaming and split bookmark prepopulation

* Configuration documentation updated

* added additional build information to the Branding.plist file:
- ios-app and ios-sdk commit hash
- copy the Branding.plist with additional app and build information as file name

* [feature/biometrical-mdm-setting ]Suggest Biometrical Unlock (#1046)

* Suggest biometrical unlock, after setting up passcode

* - do not show Cancel button, if password is enforced
- update biometrical UI switch value in settings view
- fixed bug, dismissing the view, if biometrical is not available

* moved extension into ownCloudAppShared folder

* Branding: (#1045)

- add new `branding.user-defaults-default-values` class settings key to allow registration of alternative defaults for user defaults

* - fix bug where a quick access entry showed no items when selected a second time. (#1044)

* used better wording for "Cellular transfers"

* - adjust to final capabilities naming for infinite PROPFIND support

* upgraded app version numbers

Co-authored-by: felix-schwarz <felix-schwarz@users.noreply.github.com>
Co-authored-by: Matthias Hühne <github@hosy.de>
Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

* [feature/poll-efficiency] Poll for changes efficiency enhancements (#1043)

* - change text in the footer below private links in accordance with https://github.com/owncloud/enterprise/issues/4726

* - SDK update

* - ClientRootViewController: start OCFileProviderServiceStandby only
	- after the connection status is online (== authentication successful)
	- OR 5 seconds after receiving the core from the Core Manager
- fix theoretical bug in ClientRootViewController where OCFileProviderServiceStandby could be started and not stopped if Core Manager returned a core with error

* - bump version to 11.7.2 / build 202

* - FileProviderExtension:
	- handle case where OCDatabase is not available or not open (previously led to a hang/crash as the FP waited for a DB call to return)
	- disable malloc stack logging as it leads to immediate crashes when debugging the file provider

* - update SDK to new version with support for coordinated scan for changes

* - bump build number to 203

Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

* Configuration documentation updated

* [fix/media-playback-url] Fix WebDAV endpoint URL for media playback after restoration (#1093)

* - fix https://github.com/owncloud/enterprise/issues/4924 via SDK update

* Configuration documentation updated

Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>
Co-authored-by: hosy <hosy@users.noreply.github.com>

* [feature/webfinger] Webfinger / server location (#1059)

* - change text in the footer below private links in accordance with https://github.com/owncloud/enterprise/issues/4726

* - SDK update

* added additional build information to the Branding.plist file:
- ios-app and ios-sdk commit hash
- copy the Branding.plist with additional app and build information as file name

* [feature/biometrical-mdm-setting ]Suggest Biometrical Unlock (#1046)

* Suggest biometrical unlock, after setting up passcode

* - do not show Cancel button, if password is enforced
- update biometrical UI switch value in settings view
- fixed bug, dismissing the view, if biometrical is not available

* moved extension into ownCloudAppShared folder

* Branding: (#1045)

- add new `branding.user-defaults-default-values` class settings key to allow registration of alternative defaults for user defaults

* - fix bug where a quick access entry showed no items when selected a second time. (#1044)

* used better wording for "Cellular transfers"

* - in branded clients: removed UISwitch to enable/disable deletion of available offline files in manage screen
- fixed text center layout in table view cell

* fixed UI issue in sharing view

* #4801 "Log out"  text changes

* #4801 text changes for "Prevent gestures"

* - ThemeTableViewCell, StaticTableViewRow: new .plain message style (#1053)

- add new view controller to present license texts for each component individually

* [feature/modular-localization] Modular localization (#1054)

* - Branding class:
	- move logic for VendorServices.appName to Branding.appDisplayName
	- change app.name variable value to that of Branding.appDisplayName
- switch String.localized from NSLocalizedString() to OCLocale.localize()
- SwiftLint: adapt to latest SwiftLint version

* - add .localized(replacements) function to provide easy access to custom variables support in OCLocale

* - update SDK

* [feature/mdm-auto-connect] Skip Account Screen via MDM (#1056)

* #4801 Skip "Manage" screen / automatically open "Files" screen after login via MDM parameter account.auto-connect

* - SceneDelegate: use .first (which returns nil in case there are no bookmarks) instead of bookmark(at: 0) (which crashes in case there are no bookmarks because the array is empty and the method tries to access an item at index 0)

* added auto-connect after every setup, not only the first setup

Co-authored-by: Felix Schwarz <fs-git@iospirit.com>

* [feature/mdm-biometrical-unlock]  MDM setting for Biometrical Unlock (#1058)

* #4818 control via MDM to auto enable biometrical unlock

* overwrites user defaults with MDM settings and use existing getter/setter; renamed MDM key to use-biometrical-unlock

* - use user defaults for use-biometrical-unlock
- show Face ID authorization immediately after enabling Face ID

* - StaticLoginSetupViewController: add support for OCServerLocator, requesting a username if it is enabled
- add additional debug options to ownCloud.xcscheme
- update SDK to gain access to OCServerLocator

* updated APP_VERSION number

* reverted merge conflict changes

* using correct SDK commit

* using latest development sdk commit

Co-authored-by: Matthias Hühne <github@hosy.de>
Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

* changed app version number

* added missing translation string

* [feature/rename-accounts] Rename Account without Authentication Flow (#1097)

* #972 rename account name only (bookmark) is possible without re-authentication flow

* added changelog entry

* moved duplicated code into a function for reuse

* fixed issue link in changelog entry

* Calens changelog updated

* fixed code review finding

* fixed QA findings 1,2,3
- show UI to delete authentication data in edit mode
- only proceed without authentication, if authentication data is not nil

* show delete authentication data UI in all edit mode calls

Co-authored-by: hosy <hosy@users.noreply.github.com>

* #5008 (2) only show "Settings" toolbar, if no account is configured, otherwise "Settings" would appear twice

* [feature/biometrical-button] Biometrical Authentication Button (#1098)

* #1004 added biometrical unlock button, which appears by default as fallback, if biometrical unlock does not work (in file provider or not detected)

* - added fallback image for face-id too
- added changelog entry

* Calens changelog updated

* - fixed visibility of biometrical button in edit mode and, if feature is not enabled
- removed unneeded code

* - AppLockManager: add new property biometricCancelLabel to provide a label for the Cancel-option when presenting biometric authentication
- FileProviderUI / DocumentActionViewController: use new property to display "Cancel" instead of "Enter code" for biometric authentication

Co-authored-by: hosy <hosy@users.noreply.github.com>
Co-authored-by: Felix Schwarz <fs-git@iospirit.com>

* Calens changelog updated

* removed branding test values

* Transifex translation updates (#994)

* disabled beta warning, beta build

* - added accessibility label to biometrical unlock button
- added release notes
- added missing changelog files, moved to release folder

* Calens changelog updated

* added fastlane release notes

* - update SDK to include fix/token-timetolerance changes from the SDK

* changed folder name to release date

* Calens changelog updated

* - added missing changelog entry
- changed enterprise issue to ios-app pr

* Calens changelog updated

* fixed duplicated app_version number on TestFlight

Co-authored-by: Felix Schwarz <fs-git@iospirit.com>
Co-authored-by: felix-schwarz <felix-schwarz@users.noreply.github.com>
Co-authored-by: hosy <hosy@users.noreply.github.com>
Co-authored-by: Matthias Hühne <>

felix-schwarz committed:
- update SDK to include latest Drive and Data Source changes
- update schemes to Xcode 13.3

felix-schwarz committed:
[milestone/11.9.1] Milestone 11.9.1 (#1111)

* - BookmarkViewController: re-add infinite PROPFIND code previously removed due to a merge error
- update CHANGELOG.md
- bump build and version numbers
- update ios-sdk to increase timeout for start of response to infinite PROPFIND from 30 seconds to 6 minutes

* Calens changelog updated

* [feature/fastlane-app-group-identifier] Set App Group Identifier by Fastlane via Branding.plist (#1099)

* new version number 11.9.0, enabled beta build and warning

* [feature/infinite-propfind] Infinite PROPFIND support (#1002)

* - ProgressIndicatorViewController : extend with Progress object and cancellation support / monitoring
- BookmarkViewController: add infinite PROPFIND prepopulation step after account creation
- update SDK to provide infinite PROPFIND support

* Configuration documentation updated

* - update SDK to gain access to streaming infinite PROPFIND
- switch from receive+parse to stream+parse for infinite PROPFIND

* - update SDK to gain access to new supportsInfinitePropfind capability
- BookmarkViewController: make prepopulation configurable, using new capability and new class setting
- ProgressIndicatorViewController: add support for new title label view

* - BookmarkViewController: fix property name typo
- scheme: add new options to test streaming and split bookmark prepopulation

* Configuration documentation updated

* added additional build information to the Branding.plist file:
- ios-app and ios-sdk commit hash
- copy the Branding.plist with additional app and build information as file name

* [feature/biometrical-mdm-setting ]Suggest Biometrical Unlock (#1046)

* Suggest biometrical unlock, after setting up passcode

* - do not show Cancel button, if password is enforced
- update biometrical UI switch value in settings view
- fixed bug, dismissing the view, if biometrical is not available

* moved extension into ownCloudAppShared folder

* Branding: (#1045)

- add new `branding.user-defaults-default-values` class settings key to allow registration of alternative defaults for user defaults

* - fix bug where a quick access entry showed no items when selected a second time. (#1044)

* used better wording for "Cellular transfers"

* - adjust to final capabilities naming for infinite PROPFIND support

* upgraded app version numbers

Co-authored-by: felix-schwarz <felix-schwarz@users.noreply.github.com>
Co-authored-by: Matthias Hühne <github@hosy.de>
Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

* [feature/poll-efficiency] Poll for changes efficiency enhancements (#1043)

* - change text in the footer below private links in accordance with https://github.com/owncloud/enterprise/issues/4726

* - SDK update

* - ClientRootViewController: start OCFileProviderServiceStandby only
	- after the connection status is online (== authentication successful)
	- OR 5 seconds after receiving the core from the Core Manager
- fix theoretical bug in ClientRootViewController where OCFileProviderServiceStandby could be started and not stopped if Core Manager returned a core with error

* - bump version to 11.7.2 / build 202

* - FileProviderExtension:
	- handle case where OCDatabase is not available or not open (previously led to a hang/crash as the FP waited for a DB call to return)
	- disable malloc stack logging as it leads to immediate crashes when debugging the file provider

* - update SDK to new version with support for coordinated scan for changes

* - bump build number to 203

Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

* Configuration documentation updated

* [fix/media-playback-url] Fix WebDAV endpoint URL for media playback after restoration (#1093)

* - fix https://github.com/owncloud/enterprise/issues/4924 via SDK update

* Configuration documentation updated

Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>
Co-authored-by: hosy <hosy@users.noreply.github.com>

* [feature/webfinger] Webfinger / server location (#1059)

* - change text in the footer below private links in accordance with https://github.com/owncloud/enterprise/issues/4726

* - SDK update

* added additional build information to the Branding.plist file:
- ios-app and ios-sdk commit hash
- copy the Branding.plist with additional app and build information as file name

* [feature/biometrical-mdm-setting ]Suggest Biometrical Unlock (#1046)

* Suggest biometrical unlock, after setting up passcode

* - do not show Cancel button, if password is enforced
- update biometrical UI switch value in settings view
- fixed bug, dismissing the view, if biometrical is not available

* moved extension into ownCloudAppShared folder

* Branding: (#1045)

- add new `branding.user-defaults-default-values` class settings key to allow registration of alternative defaults for user defaults

* - fix bug where a quick access entry showed no items when selected a second time. (#1044)

* used better wording for "Cellular transfers"

* - in branded clients: removed UISwitch to enable/disable deletion of available offline files in manage screen
- fixed text center layout in table view cell

* fixed UI issue in sharing view

* #4801 "Log out"  text changes

* #4801 text changes for "Prevent gestures"

* - ThemeTableViewCell, StaticTableViewRow: new .plain message style (#1053)

- add new view controller to present license texts for each component individually

* [feature/modular-localization] Modular localization (#1054)

* - Branding class:
	- move logic for VendorServices.appName to Branding.appDisplayName
	- change app.name variable value to that of Branding.appDisplayName
- switch String.localized from NSLocalizedString() to OCLocale.localize()
- SwiftLint: adapt to latest SwiftLint version

* - add .localized(replacements) function to provide easy access to custom variables support in OCLocale

* - update SDK

* [feature/mdm-auto-connect] Skip Account Screen via MDM (#1056)

* #4801 Skip "Manage" screen / automatically open "Files" screen after login via MDM parameter account.auto-connect

* - SceneDelegate: use .first (which returns nil in case there are no bookmarks) instead of bookmark(at: 0) (which crashes in case there are no bookmarks because the array is empty and the method tries to access an item at index 0)

* added auto-connect after every setup, not only the first setup

Co-authored-by: Felix Schwarz <fs-git@iospirit.com>

* [feature/mdm-biometrical-unlock]  MDM setting for Biometrical Unlock (#1058)

* #4818 control via MDM to auto enable biometrical unlock

* overwrites user defaults with MDM settings and use existing getter/setter; renamed MDM key to use-biometrical-unlock

* - use user defaults for use-biometrical-unlock
- show Face ID authorization immediately after enabling Face ID

* - StaticLoginSetupViewController: add support for OCServerLocator, requesting a username if it is enabled
- add additional debug options to ownCloud.xcscheme
- update SDK to gain access to OCServerLocator

* updated APP_VERSION number

* reverted merge conflict changes

* using correct SDK commit

* using latest development sdk commit

Co-authored-by: Matthias Hühne <github@hosy.de>
Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>

* changed app version number

* #4918 added the possiblity, to set a custom app group identifier via Branding.plist parameter "build.app-group-identifier", which will be set by fastlane to all needed key/value pairs. This is needed, if customer using a own resigning script, which does not handle setting the app group identifier.

* added documentation and changelog entry

* Calens changelog updated

* - add additional parameter build.oc-app-group-identifier that only changes the OCAppGroupIdentifier and OCKeychainAccessGroupIdentifier keys in Info.plist

* - add ownCloud/Resources/Info.plist to list of files to replace OCAppGroupIdentifier and OCKeychainAccessGroupIdentifier in

* added documentation for the new build key

* added missing translation string

* [feature/rename-accounts] Rename Account without Authentication Flow (#1097)

* #972 rename account name only (bookmark) is possible without re-authentication flow

* added changelog entry

* moved duplicated code into a function for reuse

* fixed issue link in changelog entry

* Calens changelog updated

* fixed code review finding

* fixed QA findings 1,2,3
- show UI to delete authentication data in edit mode
- only proceed without authentication, if authentication data is not nil

* show delete authentication data UI in all edit mode calls

Co-authored-by: hosy <hosy@users.noreply.github.com>

* #5008 (2) only show "Settings" toolbar, if no account is configured, otherwise "Settings" would appear twice

* [feature/biometrical-button] Biometrical Authentication Button (#1098)

* #1004 added biometrical unlock button, which appears by default as fallback, if biometrical unlock does not work (in file provider or not detected)

* - added fallback image for face-id too
- added changelog entry

* Calens changelog updated

* - fixed visibility of biometrical button in edit mode and, if feature is not enabled
- removed unneeded code

* - AppLockManager: add new property biometricCancelLabel to provide a label for the Cancel-option when presenting biometric authentication
- FileProviderUI / DocumentActionViewController: use new property to display "Cancel" instead of "Enter code" for biometric authentication

Co-authored-by: hosy <hosy@users.noreply.github.com>
Co-authored-by: Felix Schwarz <fs-git@iospirit.com>

* Calens changelog updated

* Calens changelog updated

* Calens changelog updated

* fixed merge conflicts

* Calens changelog updated

* fixed syntax error

* fixed syntax error

* added missing changelog entry

* Calens changelog updated

Co-authored-by: Felix Schwarz <fs-git@iospirit.com>
Co-authored-by: felix-schwarz <felix-schwarz@users.noreply.github.com>
Co-authored-by: hosy <hosy@users.noreply.github.com>
Co-authored-by: Matthias Hühne <>

* updated changelog and release notes

* Calens changelog updated

* - SDK update to improve handling of token refresh "responses" while offline
- add scheme env variable to turn on/off auth race condition simulator

* - ios-sdk update: fix dead-lock
- BookmarkViewController: fix finding (1) in #1109

* - update commit to latest develop commit including OCBookmarkManager deadlock fix

* - change SDK to latest develop version

* [fix/biometrical-unlock-setup] Fix: Setup Passcode with Biometrical Unlock (#1112)

* #5086 fix for passcode setup with biometrical unlock enabled:
- biometrical unlock button no longer appear in setup view
- after passcode was successfully setup, show biometrical unlock for permissions dialog

* removed unneeded comment

* added changelog and release notes

* Calens changelog updated

* reverted to Xcode 12 version

* - update to SDK release 11.9.1 (master)

* changed folder name for changelog release

* Calens changelog updated

Co-authored-by: felix-schwarz <felix-schwarz@users.noreply.github.com>
Co-authored-by: Matthias Hühne <mhuehne@owncloud.com>
Co-authored-by: hosy <hosy@users.noreply.github.com>
Co-authored-by: Matthias Hühne <>

Commit:
updated app version number because of redundant AppStore build

Commit:
[docs/folder-structure] Changed Docs Structure / MDM Fixes (#1094)

* #1070 #1082
- added missing link in MDM description
- moved ios_mdm_tables.adoc output into new folder structure
- removed no longer needed docs folder

* - moved fastlane screenshots into doc folder
- moved gomplate branding template to tools
- deleted no longer needed documentation files (values are available in configuration.adoc)
- renamed files
- deleted antora folder
- added readme file in doc folder

Co-authored-by: Matthias Hühne <>

Commit:
set up a new version 11.10.0

felix-schwarz committed:
- progress snapshot, before refactoring DataSourceCollectionViewController

felix-schwarz committed:
- CollectionViewController: generalized collection view controller based on one data source / section (WIP)

felix-schwarz committed:
- CollectionView: seperate classes in their own source files, clean up structure

Commit:
[feature/openssl-1.1.1] Migration to OpenSSL 1.1.0 (#1116)

* - migrated code to new OpenSSL 1.1.1 API
- using OpenSSL Package

* - update SDK

* - update ios-sdk

Co-authored-by: Matthias Hühne <>
Co-authored-by: Felix Schwarz <fs-git@iospirit.com>

felix-schwarz committed:
- update SDK for improved error handling
- add environment variable switch to enable/disable action-timeout-simulator host simulator

felix-schwarz committed:
- CollectionView: make progress on modular structure and implementation, adding CollectionViewCellConfiguration to pass additional parameters alongside simple objects
- update SDK
- update project to use latest OpenSSL version, following merge of SDK with ios-sdk/develop

felix-schwarz committed:
Merge branch 'milestone/11.10.0' (at 'fix/copy-timeout') into branch

felix-schwarz committed:
- SDK update for new functionality
- OCImage+ViewProvider: add support for content mode context attributes
- new collection view cell types
	- DriveListCell: list cell representing drives
	- DriveHeaderCell: subclass of DriveListCell with different layout, for representing drives as headers (with large image and text)
	- ExpandableResourceCell: shows resources at full width, allowing expansion/collapse if the resource's height exceeds a certain threshold
	- ItemListCell: representing an OCItem
- CollectionViewCellConfiguration:
	- add .core property
	- add CollectionViewCellStyle type and property to pass styling information to cells
- CollectionViewCellProvider+StandardImplementations: move cell provider registration code into specific cell classes
- CollectionViewController: allow control over list appearance
- CollectionViewSection: add support for section-wide and per-cell CollectionViewCellStyle configuration
- ClientItemViewController: new view controller to form basis of new collection view based file list UI
- refactor NSMutableAttributedString styling code from ClientItemCell.swift into NSMutableAttributedString+AppendStyled.swift
- UIFont+Weight: extension providing preferred system fonts at specific weights
- UILabel+Extension: factor out code to make labels wrap text into multiple lines (makeLabelWrapText)
- OCResourceText+ViewProvider: add view provider for OCResourceText, supporting markdown (via Down), otherwise falling back to plain text
- add Down Markdown library LICENSE into ownCloudAppShared and registration as license extension

felix-schwarz committed:
- ClientItemViewController+ItemActions: provide standard implementations for MoreItemHandling and OpenItemHandling for ClientItemViewController
- CollectionViewCellConfiguration: add new item style passing along ItemCellActionHandlers
- ItemCellActionHandlers: new class to encapsulate the different handlers for an item
- ClientItemViewController: provide ItemCellActionHandlers and make the most used item actions available

felix-schwarz committed:
- address code review findings by @hosy in https://github.com/owncloud/ios-app/pull/1092

felix-schwarz committed:
- switch to new UTType APIs
- fix most deprecation warnings
- FileProvider: replace .typeIdentifier with .contentType
- new class ClientContext:
	- used to encapsulate and pass around all important API / UI objects to all parts of the UI
	- allow customization for specific purposes through primitive "inheritance"
	- can be passed around "strongly" while storing OCCore reference "weakly" (structural barrier to accidential retains)
- replace ItemCellActionHandlers with ClientContext
	- create updated "Action" versions of MoreItemHandling, OpenItemHandling, RevealItemAction, rename InlineMessageSupport to InlineMessageCenter
	- restructure code to move MoreItemAction, OpenItemAction and InlineMessageCenter to the ClientRootViewController level (ClientRootViewController+ItemActions)
- add theming support and shadow effect to ExpandableResourceCell (via new GradientView)
- ItemListCell, CollectionViewCellConfiguration, event handling: rewire for ClientContext
- Theme: replace WeakThemeable with NSHashTable for more speed / memory efficiency
- OCResourceText+OCViewProvider: implement ThemeableTextView to add theming support
- update minimum version of PLCrashReporter to 1.10.1 (from 1.7.0)
- update SDK

felix-schwarz committed:
- ownCloudTests: fix compilation issues due to API changes

felix-schwarz committed:
- migrate app to Swift 5
- fix indentation in several places

felix-schwarz committed:
- update SDK
- squash more warnings

felix-schwarz committed:
- ThemeWindow: add static property .frontmostThemeWindow
- OCResourceText+ViewProvider.swift: move from ownCloudAppShared to app to avoid Down linker warning
- AppStatistics: switch to new API for requesting a review

felix-schwarz committed:
- add KNOWN_ISSUES.md to keep track of user-visible known issues alongside advances in the code
- FileProvider:
	- FileProviderContentEnumerator: implement an NSFileProviderEnumerator driven by OCVFSContent
	- FileProviderExtension: adapt implementation to be based on OCVFS
	- OCVFSNode+FileProviderItem: make OCVFSNode comply to NSFileProviderItem
	- first basically functioning implementation with Spaces support
- update SDK for improved VFS support

felix-schwarz committed:
- File Provider: make sure actions work with OCItems, not OCVFSNodes (and crash)

felix-schwarz committed:
- ClientContext
	- add makeActionProgressHandler() part of the MoreItemAction protocol
	- add new ContextMenuProvider protocol and property
- ClientItemViewController
	- add plus button and folder actions to navigation bar
	- add support for DisplaySettings (filtering out .space folder by default)
	- add support for item context menus
- ClientRootViewController
	- implement support for ContextMenuProvider
- CollectionViewController
	- add method to standardize resolution of indexPath -> OCDataItem
	- add provideContextMenuConfiguration method for subclassing
- update known issues

felix-schwarz committed:
- ClientContext: add .originatingViewController property to pass the view controller from which f.ex. an action originates
- ClientItemViewController: add support for .originatingViewController
- ClientRootViewController+ItemActions: use .originatingViewController instead of (wrong) self for actions (fixes lack of feedback when Copying)
- update known issues

Commit:
[milestone/11.10.0] Milestone 11.10.0 (#1117)

* set up a new version 11.10.0

* [feature/openssl-1.1.1] Migration to OpenSSL 1.1.0 (#1116)

* - migrated code to new OpenSSL 1.1.1 API
- using OpenSSL Package

* - update SDK

* - update ios-sdk

Co-authored-by: Matthias Hühne <>
Co-authored-by: Felix Schwarz <fs-git@iospirit.com>

* #1114 If app is build as EMM version, the app shows an licensing error, when running shortcut intents (could not read app bundle identifier in extension) (#1122)

Co-authored-by: Matthias Hühne <>

* [feature/iOS-15-SDK] Build with iOS 15 SDK (#1121)

* #1066 support for iOS 15 SDK: Solving UI issues, removing unneeded code

* - updated Xcode version for fastlane
- fastlane readme update

Co-authored-by: Matthias Hühne <>

* #1123 Account options were not dynamically provided for action "Delete Account Item", because of wrong intent configuration (#1125)

Co-authored-by: Matthias Hühne <>

* [fix/copy-timeout] Fix copy timeout (#1119)

* - update SDK for improved error handling
- add environment variable switch to enable/disable action-timeout-simulator host simulator

* - update SDK with changes addressing code review findings by @hosy in #1119

* - added release notes changelog files
- added in-app release notes
- changed app version number
- removed beta build warning

* added OpenSSL changelog entry

* updated Xcode version

* only perform table updates, when query is final to prevent flickering and showing empty folder message (QA finding 1, 2)

* moved changelog folder to new release folder

* updated fastlane release notes for App Store

* - change SDK reference from develop to release commit

* fixed typo

Co-authored-by: Matthias Hühne <>
Co-authored-by: Felix Schwarz <fs-git@iospirit.com>

Commit:
fixed link in changelog files, which causes that calens did not create changelog files

Commit:
[fix/changelog-11.10.0] Fixed Changelog Date 11.10.0 (#1126)

* changed changelog date

* Calens changelog updated

Co-authored-by: Matthias Hühne <>
Co-authored-by: hosy <hosy@users.noreply.github.com>

felix-schwarz committed:
- ownCloudAppFramework:
	- implement VFSManager to manage and update OCVFSCores for use throughout the app and file provider
	- implement OCVault categories to perform ID conversion using VFSManager
- File Provider:
	- update OCVFSNode and OCItem NSFileProviderItem conformance to build on the classes' OCVFSItem conformance and support OCVFSItemIDRoot
	- adopt VFSManager and support OCVFSItemIDRoot
	- fix ID mapping bug for "existing items" retrieved from cache individually by adding the bookmarkUUID to the retrieved OCItem	- FileProviderEnumeratorObserver: add support for enumeration completion
	- FileProviderContentEnumerator: modernized rebuild around OCVFSContent and async sequential queues
- update SDK

felix-schwarz committed:
- File Provider: disallow deletion, renaming and moving of a space's root folder
- update known issues

felix-schwarz committed:
- bump version to 219

felix-schwarz committed:
- FileProvider:
	- file uploads fixed (serverside)
- Photo uploads:
	- refactor to add OCLocation / drive support
	- photo upload action now correctly uploads to the respective drive (not always to the personal folder)
- bump build number to 220

felix-schwarz committed:
- AlertView: add support for text alignment
- BookmarkViewController: disable infinite PROPFIND for drive-based accounts (for now)
- ClientItemViewController: add basic "Empty folder" message support
- ThemeableCollectionViewListCell: implement themed base class based on UICollectionViewListCell
- CollectionViewCellProvider+StandardImplementations:
	- use ThemeableCollectionViewListCell for .presentable items
	- make disclosure indicator dependant on child status for .presentable items
- CollectionViewController: add support for theming the background
- DriveListCell, DriveHeaderCell, ItemListCell: inherit from ThemeableCollectionViewListCell
- ItemListCell: add themeing support
- NSObject+ThemeApplication:
	- fix appearance issue for UITabBarAppearance in iOS 15
	- add support for UICollectionViewListCell
- KNOWN_ISSUES: update, amend with ideas for further evolution, beyond known issues

felix-schwarz committed:
- bump app version to 221

felix-schwarz committed:
- add OCDataItem+InteractionProtocols allow OCDataItems to provide handling for:
	- selection / opening
	- swipe actions
	- context menu items
- OCDrive: implement DataItemSelectionInteraction
- OCItem: implement DataItemSelectionInteraction, DataItemSwipeInteraction, DataItemContextMenuInteraction, refactored from ClientRootViewController, CollectionViewProvider and ClientItemViewController
- Action: add new .emptyFolder location
- ClientItemViewController:
	- implement empty folder handling, showing actions matching to the .emptyFolder location
	- add search related code (WIP)
- ClientContext:
	- factor out ActionProgressHandler generation into ActionProgressHandlerProvider protocol
	- factor out Item Viewing into ViewItemAction protocol
	- change OpenItemAction, MoreItemAction, RevealItemAction, ContextMenuProvider to accept OCDataItem instead of OCItem
	- add new SwipeActionsProvider to provide UISwipeActionsConfiguration for trailing and leading swipes
	- add new ClientItemInteraction enum to describe all common item interactions
	- add .permissions and .permissionHandler to drive .validate(permission:for:) to validate if an item interaction is allowed
- CollectionView:
	- add section-based layout support, simplified via CollectionViewSection.CellLayout enum with layout generator

felix-schwarz committed:
- SortBarCell: WIP
- ActionCell: move OCAction+DataItemSelectionInteraction into a separate source file and update it for the latest DataItemSelectionInteraction protocol version

felix-schwarz committed:
- CollectionView:
	- add ThemeableCollectionViewCell as regular twin of ThemeableCollectionViewListCell
	- add configuredConstraints property to make changing constraints simple
	- CollectionViewCellConfiguration:
		- add method .configureCell() to simplify fetching of items to configure cell contents
		- refactor existing implementations to remove duplicate code
	- CollectionViewController:
		- add variant to use stack view as root
		- add support to add and remove "stacked" child view controllers, enabling permanent messages, toolbars and more
		- fix retain cycle in UICollectionViewCompositionalLayout(sectionProvider:)
		- add new property .animateDifferences to disable animations on view controller level
		- add new method to retrieve the items at multiple index paths at once
		- add multi-select support, including new .multiselection interaction
	- CollectionViewSection:
		- extend CellLayout.list
		- fully implement CellLayout.fullWidth
		- add new CellLayout.sideways
		- allow fully custom layouts with CellLayout.custom
		- add new property .animateDifferences to disable animations on section level
- ClientItemViewController:
	- add additional leadIn & trailing data sources for more flexible layout
	- support .loading, .empty and .hasContent ContentStates with distinct messages and different contents presented to the user
	- add SortBar and sorting support
	- add general-purpose actions bar mechanism
	- implement multi-select support using that mechanism
- ClientContext:
	- add new ClientItemInteraction.multiSelection
	- add new item-unspecific hasPermission(for: interaction) method
- SortBar:
	- show context menu instead of custom table view on top (disposing of SortMethodTableViewController)
	- remove segmented control "wide" alternative
- BookmarkViewController:
	- ensure that, for token-based auth methods, the correct name of the auth method is shown (previously always "OAuth2")
	- use localized replacements for app and auth method name
- Client Actions:
	- ensure the returned icons are rendered as templates
	- replace extension location .toolbar with .multiSelection
- ActionContext: fix bug in remove(item:) where deleteableItems was used instead of moveableItems, producing wrong results when added and removing items
- Action: add additional options to provideOCAction() for more fluid UI updates and additional completion handlers
- ActionCell: support vertical and horizontal layout, base on ThemeableCollectionViewCell
- OCViewHost: add OCViewHostContentStatus contentStatus to give a broad indication of what is shown
- ViewCell: new cell type to embed arbitrary views
- RoundedInfoView: update label text when setting .infoText
- code hygiene: replace $0 with actual var names in many places, to make the code more readable
- update SDK

jesmrec commented:
### (1) (improvement)

1. Login in an oCIS account with no spaces available

Current: List of available spaces is displayed. Only the personal account is there.
Suggestion: in case there is only one item in the list, jump directly to the file list, as in regular accounts

iPhoneXR iOS15.4.1
`95a3bfd`

jesmrec commented:
### (4)

This report contains several behaviours, all of them reproducible when a member with `Viewer` permission opens a space. I will gather all together so that all of them starts from the mentioned scenario. 

### 4.1 

If `Viewer` opens empty folder, it looks like:

<img width="340" alt="Screenshot 2022-06-14 at 17 56 51" src="https://user-images.githubusercontent.com/14894746/173622480-67e1c236-f1ee-4aa5-9b46-efabb263c20e.png">

Menu to add content is hidden because no actions are available for `Viewer`. In that case, the header should be also hidden,  showing an placeholder like the one used before (A folder image with a title or similar)

### 4.2

If `Viewer` clicks on the `+` button on the top, nothing happens (`Viewer` is not allowed to add new content)

I'd hide the `+` or show a warning about this issue (first one better IMO)

### 4.3 

I know that sharing is still a restriction, but i will comment this one:

If `Viewer` clicks on three dot button on the top, `Searching Shares...` and the spinner are there forever. Not sure if the best here is hiding the option, or there is any other bad thing happening under the hood. No infinite requests are sent to the server indeed.

### 4.4

Favorite/Unfavorite star returns non translated error for `Viewers`:

<img width="328" alt="Screenshot 2022-06-14 at 18 08 17" src="https://user-images.githubusercontent.com/14894746/173624581-43d174f8-fac0-482e-8eb9-322eb688e2d7.png">

jesmrec commented:
### (5) (improvement)

I miss the footer file/folder counter available in personal drives. Very useful!

felix-schwarz committed:
- ClientContext:
	- change .initialRootItem to .rootItem and change type from OCItem to OCDataItem
	- add new interaction types drag and acceptDrop
- OCDataItem+InteractionProtocols:
	- add new protocol DataItemDragInteraction
	- add new protocol DataItemDropInteraction
- OCItem+Interactions: implement support for new drag and drop interactions
- CollectionViewController
	- add new method section(at:) to get the CollectionViewSection for an index
	- add new method targetedDataItem(for:,interaction:) that allows modifying the item targeted by an interaction and that returns the context's rootItem if no index path is given
- CollectionViewController+DragDropSupport: implement delegates for dragging and dropping, using DataItemDragInteraction and DataItemDropInteraction
- ClientItemViewController:
	- add support for ClientContext.rootItem in combination with OCQuerys
	- subclass targetedDataItem(for:,interaction:) to return root item as target for drops for contents that actually represents the root folder
- bump version to 224
- update SDK

felix-schwarz committed:
- Action: add new .dropAction location and add relevant actions there
- ActionCell: switch from selectedBackgroundView to backgroundConfiguration and add support for drop target highlighting
- DriveHeaderCell:
	- increase contrast of background view
	- make header cell slimmer if no background image is set
- ClientContext:
	- add DropTargetsProvider that can provide drop targets when items are dragged onto a view
	- add support for multiple permission handlers
- CollectionViewController:
	- move drag and drop code from CollectionViewController+DragDropSupport into CollectionViewController to allow subclassing
	- move actions bar support from ClientItemViewController into CollectionViewController
	- add support for DropTargetsProvider
- ClientItemViewController
	- add permission handler to prevent drags and context-menus when in multi-selection mode
	- remove search code to re-add it later
- OCAction+Interactions: add drop support to trigger action run
- OCItem+Interactions: adapt DataItemDropInteraction to prevent drops of items onto themselves or their existing location

felix-schwarz committed:
- Search
	- datasource-based approach
	- SearchViewController provides and manages the UI around search
	- SearchScope:
		- implements the actual search
		- provides results as datasource
		- can reuse existing OCQuery or its own
		- can provide its own cell style for search results
- SortBar: rename SearchScope to SortBarSearchScope (pending removal)
- SortDescriptor: encapsulate both SortMethod and SortDirection into a single, extensible object
- ItemListCell:
	- improve support for showing/hiding more and reveal buttons
	- add extension to CollectionViewCellStyle for setting more/reveal button status
- CollectionViewCellStyle:
	- refactor into a class that can
		- store options and type
		- allows basic cloning with modifications
- CollectionViewController:
	- add item highlighting support
	- add support for hiding sections
	- add method for updating and reloading sections
- ClientItemViewController:
	- add support for revealing items
	- add search support
- ClientContext:
	- add support for property observation (with sample implementation for sortMethod and sortDirection) for var types that aren't KVO-observable
	- add sortDescriptor property
- OCDataItem+InteractionProtocols:
	- add optional revealItem() method to DataItemSelectionInteraction protocol

felix-schwarz committed:
- ItemListCell:
	- use backgroundConfiguration instead of backgroundView
	- turn reveal button into a proper accessory for proper layout
	- cleanup layout and animation code
	- add support for reveal highlighting
- SearchScope
	- add "Show more items" functionality to CustomQuerySearchScope

felix-schwarz committed:
- SearchScope: set result count default from 2 (debug value) to 100 (previous value)

felix-schwarz committed:
- update known issues, add new ideas regarding location picker, collection view updates
- ready for preliminary code review to merge into milestone/12.0

felix-schwarz committed:
- MessageGroupCell: change order of "Apply to all" label and switch, make the label truncating if not enough space is available (as reported by @mmattel)
- add idea for evolution of the accounts view

felix-schwarz committed:
Merge remote-tracking branch 'origin/master' into feature/graph-api

# Conflicts:
#	ios-sdk
#	ownCloud.xcodeproj/project.pbxproj
#	ownCloudAppShared/Client/User Interface/SortMethodTableViewController.swift
#	ownCloudAppShared/User Interface/Theme/NSObject+ThemeApplication.swift
#	ownCloudAppShared/User Interface/Theme/ThemeCollection.swift

hosy commented on ownCloudAppShared/Client/Actions/CreateFolderAction.swift:
```diff
@@ -106,7 +106,7 @@ open class CreateFolderAction : Action {
 
 	override open class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
-		if location == .toolbar || location == .folderAction || location == .contextMenuItem {
+		if location == .keyboardShortcut || location == .folderAction || location == .emptyFolder {
```
I would suggest:
```
if [.keyboardShortcut, .folderAction, .emptyFolder].contains(location) {
```

hosy commented on ownCloudAppShared/Client/Collection Views/Cells/MessageCell.swift:
```diff
@@ -0,0 +1,25 @@
+import ownCloudSDK
+import ownCloudApp
+
+class MessageCell : ThemeableCollectionViewListCell {
```
Does not implement code. Is it currently needed?

felix-schwarz replied:
Yes, that was a leftover from an idea I didn't yet get around to execute. I removed it now.

hosy commented on ownCloudAppShared/Client/Collection Views/View Controllers/ClientItemViewController.swift:
```diff
@@ -0,0 +1,729 @@
+
+		if let queryDatasource = query?.queryResultsDataSource {
+			emptyItemListItem = OCDataItemPresentable(reference: "_emptyItemList" as NSString, originalDataItemType: nil, version: nil)
+			emptyItemListItem?.title = "This folder is empty. Fill it with content:".localized
```
String is missing in `Localizable.strings`

felix-schwarz replied:
Thanks, I added that string to the `Localizable.strings`.

hosy commented on ownCloudAppShared/User Interface/Progress/ProgressIndicatorViewController.swift:
```diff
@@ -23,22 +23,81 @@ open class ProgressIndicatorViewController: UIViewController, Themeable {
 		progressView = UIProgressView(progressViewStyle: .bar)
 		progressView.translatesAutoresizingMaskIntoConstraints = false
 
+		if #available(iOS 13, *) {
```
iOS 13 switch can be removed

felix-schwarz replied:
I removed that code now, thanks!

hosy commented on ownCloudAppShared/UIKit Extension/UIView+OCDataItem.swift:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	public var dataItemVersion: OCDataItemVersion {
+		return NSString(format: "%p", self)
```
Can `NSString` removed by `String`?

felix-schwarz replied:
Not here unfortunately, as  only `NSString` inherits from `NSObject` and implements `NSObjectProtocol`.

hosy commented on ownCloudAppShared/UIKit Extension/UICollectionViewDiffableDataSource+Tools.swift:
```diff
@@ -0,0 +1,17 @@
+//
+//  Created by Felix Schwarz on 04.04.22.
+//  Copyright © 2022 ownCloud GmbH. All rights reserved.
+//
```
Full copyright header is missing

felix-schwarz replied:
Thanks, I added it.

hosy commented on ownCloudAppShared/UIKit Extension/LAContext+Extension.swift:
```diff
@@ -17,18 +17,38 @@
+		if  canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: nil) {
+			switch self.biometryType {
+				case .faceID : if #available(iOSApplicationExtension 13.0, *) {
+					return UIImage(systemName: "faceid")
```
iOS 13 switch is no longer needed

felix-schwarz replied:
Thanks, I fixed that.

hosy commented on ownCloudAppShared/UIKit Extension/LAContext+Extension.swift:
```diff
@@ -17,18 +17,38 @@
+					return UIImage(named: "biometrical-faceid")
+				}
+				case .touchID: if #available(iOSApplicationExtension 13.0, *) {
+					return UIImage(systemName: "touchid")
```
iOS 13 switch is no longer needed

felix-schwarz replied:
I removed the pre-iOS 13 code here as well. Thanks!

hosy reviewed (changes requested):
Please can you check, why the `docs` folder was moved.

felix-schwarz commented on KNOWN_ISSUES.md:
```diff
@@ -0,0 +1,72 @@
+#  Known issues in version 12.0 alpha 2
```
I'd keep the file around while in alpha, to quickly keep record of what's still open / not working in the respective commit. I also use(d) that list for the public alphas so far.

Alternatively, we could break those out in GitHub issues, but since there are many subtleties in there, too, I'm not sure it's worth the time.

felix-schwarz commented on ownCloud File Provider/FileProviderContentEnumerator.m:
```diff
@@ -0,0 +1,687 @@
+	- inform the observer about the items returned by the server (possibly multiple times)
+	- inform the observer that you are finished with this page
+	*/
+}
```
Those are the comments from Apple's template. Since FileProvider documentation spreads out in WWDC videos, header files, developer documentation and those comments, I'd like to keep them around for reference.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/DeleteAction.swift:
```diff
@@ -97,8 +97,8 @@ class DeleteAction : Action {
-		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem {
-			return UIImage(named: "trash")
+		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem || location == .multiSelection || location == .dropAction {
+			return UIImage(named: "trash")?.withRenderingMode(.alwaysTemplate)
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/DuplicateAction.swift:
```diff
@@ -77,8 +77,8 @@ class DuplicateAction : Action {
-		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem {
-			return UIImage(named: "duplicate-file")
+		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem || location == .multiSelection || location == .dropAction {
+			return UIImage(named: "duplicate-file")?.withRenderingMode(.alwaysTemplate)
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/MakeAvailableOfflineAction.swift:
```diff
@@ -65,7 +65,7 @@ class MakeAvailableOfflineAction: Action {
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
 		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem {
-			return UIImage(named: "available-offline")
+			return UIImage(named: "available-offline")?.withRenderingMode(.alwaysTemplate)
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/MakeUnavailableOfflineAction.swift:
```diff
@@ -81,7 +81,7 @@ class MakeUnavailableOfflineAction: Action {
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
 		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem {
-			return UIImage(named: "available-offline")
+			return UIImage(named: "available-offline")?.withRenderingMode(.alwaysTemplate)
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/MoveAction.swift:
```diff
@@ -77,8 +77,8 @@ class MoveAction : Action {
-		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem {
-			return UIImage(named: "folder")
+		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem || location == .multiSelection || location == .dropAction {
+			return UIImage(named: "folder")?.withRenderingMode(.alwaysTemplate)
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/OpenInAction.swift:
```diff
@@ -169,7 +169,7 @@ class OpenInAction: Action {
 
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
-		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem {
+		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .contextMenuItem || location == .multiSelection || location == .dropAction {
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/UnfavoriteAction.swift:
```diff
@@ -56,7 +56,7 @@ class UnfavoriteAction : Action {
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
 		if location == .moreItem || location == .moreDetailItem || location == .contextMenuItem {
-			return UIImage(named: "star")
+			return UIImage(named: "star")?.withRenderingMode(.alwaysTemplate)
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/UnshareAction.swift:
```diff
@@ -121,8 +121,8 @@ class UnshareAction : Action {
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
-		if location == .moreItem || location == .moreDetailItem || location == .moreFolder {
-			return UIImage(named: "trash")
+		if location == .moreItem || location == .moreDetailItem || location == .moreFolder || location == .multiSelection {
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/UploadFileAction.swift:
```diff
@@ -52,7 +53,7 @@ class UploadFileAction: UploadBaseAction {
 
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
-		if location == .folderAction {
+		if location == .folderAction || location == .emptyFolder {
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/Actions+Extensions/UploadMediaAction.swift:
```diff
@@ -196,7 +196,7 @@ class UploadMediaAction: UploadBaseAction {
 
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
-		if location == .folderAction {
+		if location == .folderAction || location == .emptyFolder {
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/Actions/EditDocumentViewController.swift:
```diff
@@ -146,7 +147,7 @@ class EditDocumentViewController: QLPreviewController, Themeable {
 			guard let markupButton = self.navigationItem.rightBarButtonItems?.filter({$0.customView != nil}).first?.customView as? UIButton else { return }
 			markupButton.sendActions(for: .touchUpInside)
-		}
+		} */
```
Thanks. I removed that code.

felix-schwarz commented on ownCloud/Client/Actions/Scanner/ScanAction.swift:
```diff
@@ -94,13 +94,12 @@ class ScanAction: Action, VNDocumentCameraViewControllerDelegate {
 	override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {
-		if location == .folderAction {
-			return UIImage(systemName: "doc.text.viewfinder", withConfiguration: UIImage.SymbolConfiguration(pointSize: 26, weight: .regular))
+		if location == .folderAction || location == .emptyFolder {
```
I figured we shouldn't need those `if` statements at all and removed them in all `Action` subclasses.

felix-schwarz commented on ownCloud/Client/ClientRootViewController.swift:
```diff
@@ -353,18 +372,100 @@ class ClientRootViewController: UITabBarController, BookmarkContainer, ToolAndTa
+//							}), for: core.projectDrivesDataSource)
+//						})
+
+						topLevelViewController = CollectionViewController(context: ClientContext(with: self.rootContext, navigationController: self.filesNavigationController), sections: [
```
It was early test / sample code. I removed it. Thanks!

felix-schwarz committed:
Fix code review findings from #1101.

hosy reviewed (changes requested):
Are you sure all `if`-Statements can be removed in `override class func iconForLocation(_ location: OCExtensionLocationIdentifier) -> UIImage? {` in the `Action` subclasses?

I am not sure, if sometimes should not be a `UIImage` returned. But maybe I am wrong.

felix-schwarz commented:
@hosy The `iconForLocation()` methods were only checking if the icon is requested for a location that's already listed in the action's `locations` property. If an action's `locations` property doesn't contain a location, however, it will not be returned for that location and `iconForLocation()` will not be called.

The point of providing a location to `iconForLocation()` originally was so that actions can return different icons / icon styles for different locations. Since no alternative icons are provided at this point, the location checks can be removed from `iconForLocation()` at this point.

felix-schwarz commented:
Just noticed this comment from the first code review:

> Please can you check, why the `docs` folder was moved.

I looked into it and it was removed as a result of https://github.com/owncloud/ios-app/pull/1094 being merged in the `master` branch.

hosy reviewed (changes requested):
I still see that the following folders are not correct, comparing to the `master` branch
- .github
- changelog
- doc
- docs
- enterprise/resign
- external/libzip
- fastlane
Please can you have a look. I think these folders were not correct merged into this branch.

felix-schwarz committed:
- add missing changelogs from master

felix-schwarz committed:
Calens changelog updated

felix-schwarz committed:
- bring CHANGELOG.md in sync with master

felix-schwarz committed:
Merge branch 'feature/graph-api' of https://github.com/owncloud/ios-app into feature/graph-api

felix-schwarz commented:
@hosy I compared these folders to `master` and the only differences I could find were in `CHANGELOG.md` and the `changelog` subfolder `changelog/11.10.1_2022-08-02` missing. I brought these in sync with `master`.

The see differences in these folders when comparing them to `milestone/12.0`, but those are the changes that will bring these folders in `milestone/12.0` back in sync with `master` after `feature/graph-api` has been merged into `milestone/12.0`.
